### DISCUSSION TITLE
Asp.net Core MVC and CslaModelBinderProvider?

### DISCUSSION THREAD

**Discussion by hoangcuongvn** [MarimerLLC/csla#2914] [Questions]
In Csla 5, we had to add `CslaModelBinderProvider` into MVC's `ModelBinderProviders` as the code below:
```
services.AddControllersWithViews(config => config.ModelBinderProviders.Insert(0, new Csla.Web.Mvc.CslaModelBinderProvider()));
```
Now in the latest Csla 6, the `CslaModelBinderProvider` class no longer has its empty constructor method and we cannot use the code above. Do we still need to add `CslaModelBinderProvider` in Csla 6 and how? Or is it automatically done somewhere?

**Comment by rockfordlhotka** [comment]
You probably do still need `CslaModelBinderProvider`, and will now need to get it via dependency injection.

**Comment by hoangcuongvn** [comment]
Where can I add `CslaModelBinderProvider` in Csla 6? In Csla 5, it is added in `AddControllersWithViews`. But in that method there is no way to get `CslaModelBinderProvider` through DI.

**Reply by rockfordlhotka** [reply]
Yeah, this is a conundrum. 

The model binder needs `ApplicationContext` so it can create instances of domain types. The model binder is created by the model binder provider, so the provider needs `ApplicationContext`. Which, of course, doesn't yet exist as services are being set up, and sadly there's no overload of `AddControllersWithViews` that allows lazy loading of services like there is with methods like `AddTransient`, etc.

It feels like a chicken and egg problem doesn't it? We need to create an object instance without DI, and that instance later needs access to DI-provided services.

**Reply by rockfordlhotka** [reply]
Two solutions (older and newer/better) are described in this [SO thread](https://stackoverflow.com/questions/35640858/how-do-i-use-custom-model-binder-that-supports-dependency-injection-in-asp-net-c).

**Reply by rockfordlhotka** [reply]
Done in #2920

**Reply by CROprogrammer** [reply]
What's the new way of adding CslaModelBinderProvider then?

MVC Example which uses 6.2.2 version still uses
```
services.AddControllersWithViews((c) =>
        c.ModelBinderProviders.Insert(0, new Csla.Web.Mvc.CslaModelBinderProvider()));
```
Is this still the way to go then?
